File: A2/command-python-service/src/repositories/dashboard_repository.py
from typing import Dict, List, Optional, Any
import json


class DashboardRepository:
    def __init__(self, pool):
        self._pool = pool

    async def find_all(self) -> List[Dict[str, Any]]:
        async with self._pool.acquire() as conn:
            rows = await conn.fetch('SELECT * FROM tactical_dashboards ORDER BY created_at DESC')
            return [
                {
                    'dashboard_id': row['dashboard_id'],
                    'mission_id': row['mission_id'],
                    'sensor_summary': row['sensor_summary'],
                    'threat_level': row['threat_level'],
                    'status': row['status'],
                    'created_at': row['created_at'].isoformat() + 'Z',
                    'updated_at': row['updated_at'].isoformat() + 'Z'
                }
                for row in rows
            ]

    async def find_by_id(self, dashboard_id: str) -> Optional[Dict[str, Any]]:
        async with self._pool.acquire() as conn:
            row = await conn.fetchrow(
                'SELECT * FROM tactical_dashboards WHERE dashboard_id = $1',
                dashboard_id
            )
            
            if not row:
                return None
            
            return {
                'dashboard_id': row['dashboard_id'],
                'mission_id': row['mission_id'],
                'sensor_summary': row['sensor_summary'],
                'threat_level': row['threat_level'],
                'status': row['status'],
                'created_at': row['created_at'].isoformat() + 'Z',
                'updated_at': row['updated_at'].isoformat() + 'Z'
            }

    async def create(self, dashboard_data: Dict[str, Any]) -> Dict[str, Any]:
        async with self._pool.acquire() as conn:
            row = await conn.fetchrow(
                '''INSERT INTO tactical_dashboards (dashboard_id, mission_id, sensor_summary, threat_level, status)
                   VALUES ($1, $2, $3, $4, $5)
                   RETURNING *''',
                dashboard_data['dashboard_id'],
                dashboard_data['mission_id'],
                json.dumps(dashboard_data['sensor_summary']),
                dashboard_data['threat_level'],
                dashboard_data['status']
            )
            
            return {
                'dashboard_id': row['dashboard_id'],
                'mission_id': row['mission_id'],
                'sensor_summary': row['sensor_summary'],
                'threat_level': row['threat_level'],
                'status': row['status'],
                'created_at': row['created_at'].isoformat() + 'Z',
                'updated_at': row['updated_at'].isoformat() + 'Z'
            }

    async def update(self, dashboard_id: str, updates: Dict[str, Any]) -> Dict[str, Any]:
        fields = []
        values = []
        param_count = 1

        if 'sensor_summary' in updates:
            fields.append(f'sensor_summary = ${param_count}')
            values.append(json.dumps(updates['sensor_summary']))
            param_count += 1
        if 'threat_level' in updates:
            fields.append(f'threat_level = ${param_count}')
            values.append(updates['threat_level'])
            param_count += 1
        if 'status' in updates:
            fields.append(f'status = ${param_count}')
            values.append(updates['status'])
            param_count += 1

        if not fields:
            return await self.find_by_id(dashboard_id)

        values.append(dashboard_id)
        async with self._pool.acquire() as conn:
            row = await conn.fetchrow(
                f'''UPDATE tactical_dashboards SET {', '.join(fields)} 
                    WHERE dashboard_id = ${param_count} RETURNING *''',
                *values
            )
            
            return {
                'dashboard_id': row['dashboard_id'],
                'mission_id': row['mission_id'],
                'sensor_summary': row['sensor_summary'],
                'threat_level': row['threat_level'],
                'status': row['status'],
                'created_at': row['created_at'].isoformat() + 'Z',
                'updated_at': row['updated_at'].isoformat() + 'Z'
            }

File: A4/services/node-command-service/src/repositories/dashboardRepository.js
const { getPool } = require('../db/connection');

class DashboardRepository {
  constructor(pool) {
    this.pool = pool;
  }

  async findAll() {
    const result = await this.pool.query(
      'SELECT * FROM tactical_dashboards ORDER BY created_at DESC'
    );
    return result.rows.map(row => ({
      dashboard_id: row.dashboard_id,
      mission_id: row.mission_id,
      sensor_summary: row.sensor_summary,
      threat_level: row.threat_level,
      status: row.status,
      created_at: row.created_at.toISOString(),
      updated_at: row.updated_at.toISOString()
    }));
  }

  async findById(id) {
    const result = await this.pool.query(
      'SELECT * FROM tactical_dashboards WHERE dashboard_id = $1',
      [id]
    );
    
    if (result.rows.length === 0) {
      return null;
    }
    
    const row = result.rows[0];
    return {
      dashboard_id: row.dashboard_id,
      mission_id: row.mission_id,
      sensor_summary: row.sensor_summary,
      threat_level: row.threat_level,
      status: row.status,
      created_at: row.created_at.toISOString(),
      updated_at: row.updated_at.toISOString()
    };
  }

  async create(dashboardData) {
    const { dashboard_id, mission_id, sensor_summary, threat_level, status } = dashboardData;
    
    const result = await this.pool.query(
      `INSERT INTO tactical_dashboards (dashboard_id, mission_id, sensor_summary, threat_level, status)
       VALUES ($1, $2, $3, $4, $5)
       RETURNING *`,
      [dashboard_id, mission_id, JSON.stringify(sensor_summary), threat_level, status]
    );
    
    const row = result.rows[0];
    return {
      dashboard_id: row.dashboard_id,
      mission_id: row.mission_id,
      sensor_summary: row.sensor_summary,
      threat_level: row.threat_level,
      status: row.status,
      created_at: row.created_at.toISOString(),
      updated_at: row.updated_at.toISOString()
    };
  }

  async update(id, updates) {
    const fields = [];
    const values = [];
    let paramCount = 1;

    if (updates.sensor_summary !== undefined) {
      fields.push(`sensor_summary = $${paramCount++}`);
      values.push(JSON.stringify(updates.sensor_summary));
    }
    if (updates.threat_level !== undefined) {
      fields.push(`threat_level = $${paramCount++}`);
      values.push(updates.threat_level);
    }
    if (updates.status !== undefined) {
      fields.push(`status = $${paramCount++}`);
      values.push(updates.status);
    }

    if (fields.length === 0) {
      return await this.findById(id);
    }

    values.push(id);
    const result = await this.pool.query(
      `UPDATE tactical_dashboards SET ${fields.join(', ')} WHERE dashboard_id = $${paramCount} RETURNING *`,
      values
    );

    const row = result.rows[0];
    return {
      dashboard_id: row.dashboard_id,
      mission_id: row.mission_id,
      sensor_summary: row.sensor_summary,
      threat_level: row.threat_level,
      status: row.status,
      created_at: row.created_at.toISOString(),
      updated_at: row.updated_at.toISOString()
    };
  }
}

module.exports = DashboardRepository;
